pkg/ingestor/parser/parser.go

```diff
@@ -0,0 +1,261 @@
+//
+// Copyright 2022 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+
+	"github.com/guacsec/guac/pkg/assembler"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/verifier"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+const (
+	algorithmSHA256 string = "sha256"
+)
+
+type graphBuilder struct {
+	foundIdentities       []assembler.IdentityNode
+	foundSubjectArtifacts []assembler.ArtifactNode
+	foundAttestations     []assembler.AttestationNode
+	foundBuilders         []assembler.BuilderNode
+	foundDependencies     []assembler.ArtifactNode
+}
+
+func newGraphBuilder() *graphBuilder {
+	return &graphBuilder{
+		foundIdentities:       []assembler.IdentityNode{},
+		foundSubjectArtifacts: []assembler.ArtifactNode{},
+		foundAttestations:     []assembler.AttestationNode{},
+		foundBuilders:         []assembler.BuilderNode{},
+		foundDependencies:     []assembler.ArtifactNode{},
+	}
+}
+
+// ParseDocumentTree
+func ParseDocumentTree(docTree processor.DocumentTree) (assembler.AssemblerInput, error) {
+
+	builder := newGraphBuilder()
+	err := builder.parse(docTree)
+	if err != nil {
+		return assembler.AssemblerInput{}, err
+	}
+
+	assemblerinput := assembler.AssemblerInput{
+		Nodes: builder.createNodes(),
+		Edges: builder.createEdges(),
+	}
+
+	return assemblerinput, nil
+}
+
+func (b *graphBuilder) createEdges() []assembler.GuacEdge {
```

Comment:
hmm, it looks like this is creating a complete subgraph between types of nodes. Is this intended or just a step towards what we want. I think if this is done locally per node, that would work in creating edges recursively, but it seems here that it's parsing the entire tree and creating edges between each node found within the entire tree?

Not sure if i'm missing something, maybe you can help clarify

Reply:
fixed to create assembler input per document node